Tools/refcounts/lint.py

```diff
@@ -0,0 +1,301 @@
+"""Lint Doc/data/refcounts.dat."""
+
+from __future__ import annotations
+
+import itertools
+import re
+import tomllib
+from argparse import ArgumentParser, RawDescriptionHelpFormatter
+from dataclasses import dataclass, field
+from enum import auto as _auto, Enum
+from pathlib import Path
+from typing import TYPE_CHECKING, LiteralString, NamedTuple
+
+if TYPE_CHECKING:
+    from collections.abc import Callable, Iterable, Mapping
+
+ROOT = Path(__file__).parent.parent.parent.resolve()
+DEFAULT_REFCOUNT_DAT_PATH: str = str(ROOT / 'Doc/data/refcounts.dat')
+DEFAULT_STABLE_ABI_TOML_PATH: str = str(ROOT / 'Misc/stable_abi.toml')
+
+C_ELLIPSIS: LiteralString = '...'
+
+MATCH_TODO: Callable[[str], re.Match | None]
+MATCH_TODO = re.compile(r'^#\s*TODO:\s*(\w+)$').match
+
+OBJECT_TYPES: frozenset[str] = frozenset()
+
+for qualifier, object_type, suffix in itertools.product(
+    ('const ', ''),
+    (
+        'PyObject',
+        'PyLongObject', 'PyTypeObject',
+        'PyCodeObject', 'PyFrameObject',
+        'PyModuleObject', 'PyVarObject',
+    ),
+    ('*', '**', '* const *', '* const*'),
+):
+    OBJECT_TYPES |= {
+        f'{qualifier}{object_type}{suffix}',
+        f'{qualifier}{object_type} {suffix}',
+    }
+del suffix, object_type, qualifier
+
+IGNORE_LIST: frozenset[str] = frozenset((
+    # part of the stable ABI but should not be used at all
+    'PyUnicode_GetSize',
+    # part of the stable ABI but completely removed
+    '_PyState_AddModule',
+))
+
+def flno_(lineno: int) -> str:
+    # Format the line so that users can C/C from the terminal
+    # the line number and jump with their editor using Ctrl+G.
+    return f'{lineno:>5} '
+
+class RefType(Enum):
+    UNKNOWN = _auto()
+    UNUSED = _auto()
+    DECREF = _auto()
+    BORROW = _auto()
+    INCREF = _auto()
+    STEALS = _auto()
+    NULL = _auto()  # for return values only
+
+class LineInfo(NamedTuple):
```

Comment:
I think it's worth making this a dataclass for future maintenace.

Reply:
Oops, I hit "add single comment" instead of "start a review." Sorry! Expect a more thorough review soon.

Reply:
> Everything should be read-only IMO.

We can use a `frozen` parameter to achieve this

Reply:
Well, I decided to drop the NamedTuple (I usually prefer using NamedTuple when I can since dataclasses)

Reply:
FWIW I fully agree that NamedTuples are bad for APIs that you're exposing for other users in a library, but I think they're still great for internal-only tools that you don't expect to be used by other people. They can be faster and more lightweight, and being able to easily unpack values in a tuple-like way can be a feature rather than a bug in internal API.

However, dataclasses are also great! So I'm definitely not asking for this change to be reverted 😄